### ISSUE TITLE
Resource cleanup when a transaction is closed

### ISSUE THREAD

**Issue by alexjpwalker** [graknlabs/grakn#5905]
## Problem to Solve

Resources are not necessarily being cleaned up in all the cases where a transaction is closed.

## Proposed Solution

Ensure that the following behaviours occur in the following scenarios:

Scenario | Expected behaviour |
|---|---|
| Successful commit | Send OK response, close RocksTransaction, close gRPC stream |
| Error in commit or query | Send error response, close transaction, close stream |
| Client closes stream | Close transaction |
| Error occurs when client closes stream | Server left in a consistent state |
| Error when opening transaction | Send error response, close stream |
| Session gets closed | Send error response to all its transaction streams, close the transactions, close the streams |
| Database gets deleted | Same behaviour as "Session gets closed" for all its sessions |
| Server is shut down gracefully | Same behaviour as "Session gets closed" for all sessions |
| Client dies suddenly | Send error response to all other transaction streams owned by the Session that owns this stream; close them and close the transactions |
| Server terminates suddenly | All sessions and transactions disappear and data integrity is maintained |
| Connection drops | TBC |

Of these cases, the last two need special attention:
- **Server terminates suddenly**: All sessions and transactions should disappear and data integrity should be maintained. This is a fundamental principle of ACID - we should write durability tests (the D in ACID) to validate this
- **Connection drops**: Need to decide what to do here - should we attempt to reconnect? Is that even possible with gRPC? If we simply give up on the connection then it's equivalent to the client dying suddenly.

**Comment by lolski** [comment]
[Triaged] I'm putting this issue in the backlog because the issue is major and will require design:

1. adding durability test in order to comply with ACID
2. adding a new connection recovery mechanism

**Comment by haikalpribadi** [comment]
I agree it should go to Backlog, but I think this is a high priority.
